File: fuzzy/Fuzzy.c
//Evan Gray - May 2012
//Compile: gcc Spinner.c libcAI.so -o Spinner
//Run: ./Spinner
#include "cAI.h"
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

//#define DEBUGTURN
//#define DEBUGTHRUST
#define RAD(x) x/180*atan(1)*4
//Alive run, only run when alive
#define AR(x) if(selfAlive()){x;}
#ifdef DEBUGTURN
  #define TURNDEBUG(x) x;
#else
  #define TURNDEBUG(x)
#endif

#ifdef DEBUGTHRUST
  #define THRUSTDEBUG(x) x;
#else
  #define THRUSTDEBUG(x)
#endif


//integral of ax^2+bx over x values between x=minx and x=maxx
float integrateLinearX(float a, float b, float minx, float maxx)
{
  if(minx > maxx) return 0.0f;
  else return a*(pow(maxx, 3)-pow(minx, 3))/3.0f + b*(pow(maxx, 2)-pow(minx, 2))/2.0f;
}
//integral of ax^2+bx over x values between y=0 and y=maxy
float integrateLinearXWithY(float a, float b, float maxy)
{
  float x1 = (maxy-b)/a;
  float x2 = -b/a;
  if(x1 > x2) return integrateLinearX(a, b, x2, x1);
  else return integrateLinearX(a, b, x1, x2);
}
//for area under linear function ax + b over x values between x=minx and x=maxx
float integrateLinear(float a, float b, float minx, float maxx)
{
  if(minx > maxx) return 0.0f;
  else return a*(pow(maxx, 2)-pow(minx, 2))/2.0f + b*((maxx-minx));
}
//for area under linear function ax + b over x values between y=0 and y=maxy
float integrateLinearWithY(float a, float b, float maxy)
{
  float x1 = (maxy-b)/a;
  float x2 = -b/a;
  if(x1 > x2) return integrateLinear(a, b, x2, x1);
  else return integrateLinear(a, b, x1, x2);
}
float integrateBox(float y, float minx, float maxx)
{
  if(minx > maxx) return 0.0f;
  else return y * (maxx - minx);
}
float integrateBoxX(float y, float minx, float maxx)
{
  if(minx > maxx) return 0.0f;
  else return y * (pow(maxx,2) - pow(minx,2))/2.0f;
}

//integral of piecewise function defined by f1a*x+f1b, y=maxy, f2a*x+f2b
//assume f1 is left and f3 is right
float integrateDoubleLinear(float maxy, float f1a, float f1b, float f2a, float f2b)
{
  float p1 = (maxy-f1b)/f1a;
  float p2 = (maxy-f2b)/f2a;
  //assume p2 > p1
  return integrateLinearWithY(f1a, f1b, maxy) + integrateBox(maxy, p1, p2) + integrateLinearWithY(f2a, f2b, maxy);
}

//integral piecewise function defined by (f1a*x+f1b)x, y=maxy*x, (f2a*x+f2b)x
//assume f1 is left and f3 is right
float integrateDoubleLinearX(float maxy, float f1a, float f1b, float f2a, float f2b)
{
  float p1 = (maxy-f1b)/f1a;
  float p2 = (maxy-f2b)/f2a;
  //assume p2 > p1
  return integrateLinearXWithY(f1a, f1b, maxy) + integrateBoxX(maxy, p1, p2) + integrateLinearXWithY(f2a, f2b, maxy);
}
//calculate centroid with sum of all integrateDoubleLinearX over integrateDoubleLinear

float tanAngle(int x) //converts angle to tan(x/2) so that front is 0, left is positive, right is negative
{
  if(x==180) return 50.0f;
  else return fmax(fmin(tan(RAD(x/2.0f)), 50.0f),-50.0f);
}
//these membership functions are shared between linguistic variables ClosestAngle and FurthestAngle
//anticlockwise rotation angle
float mem_Angle_Front(int x)
{
  float tanx = tanAngle(x);
  if(tanx <= 0) return fmax(1.92409*tanx+1.48335f, 0.0f);
  else return fmax(-1.92409*tanx+1.48335f, 0.0f);
}
float mem_Angle_Left(int x)
{
  float tanx = tanAngle(x);
  return fmin(fmax(4.23226*tanx-0.18154f, 0.0f), 1.0f);
}
float mem_Angle_Right(int x)
{
  float tanx = tanAngle(x);
  return fmin(fmax(-4.23226*tanx-0.18154f, 0.0f), 1.0f);
}
//distance to closest wall
float mem_Wall_Safe(int x)
{
  return fmin(fmax(0.01999f*x-5.9509f, 0.0f), 1.0f);
}
float mem_Wall_Close(int x)
{
  if(0.01382f*x-1.88270f < 1) return fmin(fmax(0.01382f*x-1.88270f, 0.0f), 1.0f);
  else return fmin(fmax(-0.01993f*x+3.002395f, 0.0f),1.0f);
}
float mem_Wall_Danger(int x)
{
  return fmin(fmax(-0.00022f*x+1.062527f, 0.0f), 1.0f);
}

float mem_Speed_Slow(float x)
{
  return fmin(fmax(-0.19666f*x+3.190827, 0.0), 1.0);
}

float mem_Speed_Medium(float x)
{
  if (2.66264f*x-3.2405f < 1.0)
  {
    return fmin(fmax(2.66264f*x-3.2405f, 0.0), 1.0);
  }
  else
  {
    return fmin(fmax(-5.55055f*x+4.795709f, 0.0), 1.0);
  }
}
float mem_Speed_Fast(float x)
{
  return fmin(fmax(2.67478f*x-3.0727f, 0.0), 1.0);
}


//centroid for turn. requires three fuzzy inputs and outputs clear value
float centroidTurn(float turnLeft, float noTurn, float turnRight)
{
  float denum = integrateDoubleLinear(turnLeft,2,0,-2,2) + integrateDoubleLinear(noTurn,3,-2,-3,4) + integrateDoubleLinear(turnRight,2,-2,-2,4);
  if(denum == 0.0f) return 1.0f;
  float num = integrateDoubleLinearX(turnLeft,2,0,-2,2) + integrateDoubleLinearX(noTurn,3,-2,-3,4) + integrateDoubleLinearX(turnRight,2,-2,-2,4);
  return num/denum;
}

float and(float a, float b)
{
  return a < b ? a : b;
}
float or(float a, float b)
{
  return a > b ? a : b;
}

//If ClosestWall Danger and closestAngle Left then turn right
float r1(float closest, float closestAngle)
{
  return and(mem_Wall_Danger(closest), mem_Angle_Left(closestAngle));
}
//If ClosestWall Danger and closestAngle Right then turn left
float r2(float closest, float closestAngle)
{
  return and(mem_Wall_Danger(closest), mem_Angle_Right(closestAngle));
}
//If ClosestWall Close and closestAngle Left then turn right
float r3(float closest, float closestAngle)
{
  return and(mem_Wall_Close(closest), mem_Angle_Left(closestAngle));
}
//If ClosestWall Close and closestAngle Right then turn left
float r4(float closest, float closestAngle)
{
  return and(mem_Wall_Close(closest), mem_Angle_Right(closestAngle));
}
//If ClosestWall Safe Noturn
float r5(float closest)
{
  return mem_Wall_Safe(closest);
}
//If FurthestAngle Right and slow turn Right
float r6(float furthestAngle, float speed)
{
  return and(mem_Angle_Right(furthestAngle), mem_Speed_Slow(speed));
}
//If FurthestAngle Left and slow turn Left
float r7(float furthestAngle, float speed)
{
  return and(mem_Angle_Left(furthestAngle), mem_Speed_Slow(speed));
}
//If FurthestAngle Forward dont turn
float r8(float furthestAngle, float speed)
{
  return and(mem_Angle_Front(furthestAngle), mem_Speed_Medium(speed));
}

//If FurthestAngle Forward and speed is medium dont turn
float r9(float furthestAngle, float speed)
{
  return and(mem_Angle_Front(furthestAngle), mem_Speed_Medium(speed));

}
//If closestAngle front and furthestAngle right, turn right
float r10(float closestAngle, float furthestAngle)
{
  return and(mem_Angle_Front(closestAngle), mem_Angle_Right(furthestAngle));
}

//If closestAngle front and furthestAngle left, turn left
float r11(float closestAngle, float furthestAngle)
{
  return and(mem_Angle_Front(closestAngle), mem_Angle_Left(furthestAngle));
}
//If furthestAngle right and speed fast, turn right
float r12(float furthestAngle, float speed)
{
  return and(mem_Speed_Fast(speed), mem_Angle_Right(furthestAngle));
}
//If furthestAngle right and speed fast, turn left
float r13(float furthestAngle, float speed)
{
  return and(mem_Speed_Fast(speed), mem_Angle_Left(furthestAngle));
}
//If FurthestAngle Right turn Right
float r14(float furthestAngle)
{
  return mem_Angle_Right(furthestAngle);
}
//If FurthestAngle Left turn Left
float r15(float furthestAngle)
{
  return mem_Angle_Left(furthestAngle);
}

float turnRules(float closest, float closestAngle, float furthestAngle, float speed)
{
  float q = mem_Wall_Safe(closest);
  float w = mem_Wall_Close(closest);
  float e = mem_Angle_Left(closestAngle);
  float r = mem_Angle_Right(closestAngle);
  //printf("safe: %.2f close: %.2f left: %.2f right: %.2f\n", q, w, e, r);

  float turnLeft  = r2(closest, closestAngle) 
                  + r4(closest, closestAngle) 
                  + r7(furthestAngle, speed)
                  + r11(closestAngle, furthestAngle)
                  + r13(furthestAngle, speed)
                  + 0.8 * r15(furthestAngle);
  float noTurn = r5(closest);
  float turnRight = r1(closest, closestAngle) 
                  + r3(closest, closestAngle) 
                  + r6(furthestAngle, speed)
                  + r10(closestAngle, furthestAngle)
                  + r12(furthestAngle, speed)
                  + 0.8 * r14(furthestAngle);

  //AR(printf("left: %.2f noturn: %.2f right: %.2f\n", turnLeft, noTurn, turnRight));
  return centroidTurn(turnLeft, noTurn, turnRight);
}

int AI_loop() {
  srand((unsigned int)time(NULL));
  setTurnSpeedDeg(20);
  int aimDir = aimdir(0);
  //Release keys
  //Set variables
  double heading = selfHeadingDeg();
  double tracking = selfTrackingDeg();
  double trackWall = wallFeeler(500,tracking);

  //clockwise rotation around the ship
  double frontWall = wallFeeler(500,heading);
  double wall1 = wallFeeler(500,heading+5);
  double wall2 = wallFeeler(500,heading+60);
  double wall3 = wallFeeler(500,heading+90);
  double wall4 = wallFeeler(500,heading+120);
  double wall5 = wallFeeler(500,heading+150);
  double backWall = wallFeeler(500,heading+180);
  double wall7 = wallFeeler(500,heading+210);
  double wall8 = wallFeeler(500,heading+240);
  double wall9 = wallFeeler(500,heading+270);
  double wall10 = wallFeeler(500,heading+300);
  double wall11 = wallFeeler(500,heading+355);
  
  int shouldThrust = 0;
  //turn right = 1, turn left = -1
  int turnDir = 0;

  

  double furthest = 0.0;
  int furthest_angle = 0;
  double closest = 600.0;
  int closest_angle = 0;
  for(int i=0;i<360;i++)
  {
    double dist = wallFeeler(1000,heading+i);
    if(dist > furthest)
    {
      furthest = dist;
      furthest_angle = i;
    }
    if(dist < closest)
    {
      closest = dist;
      closest_angle = i;
    }
  }

  int headingTrackingDiff = (int)(heading + 360 - tracking) % 360;
  int headingAimingDiff = (int)(heading + 360 - aimDir) % 360;
  int shotDanger = shotAlert(0);
  if(shotDanger > 0 && shotDanger < 200 && frontWall > 200 && trackWall > 80 && wall1 > 100 && wall11 > 100)
  {
    THRUSTDEBUG(AR(printf("avoiding shot: %d\n", shotDanger)));
    shouldThrust = 1;
  }
  else if((furthest_angle == 0) && selfSpeed() < 5 && frontWall > 200)
  {
    THRUSTDEBUG(AR(printf("propulsion\n")));
    shouldThrust = 1;
  }
  else if(headingTrackingDiff > 110 && headingTrackingDiff < 250 && trackWall < 300 && frontWall > 100)
  {
    THRUSTDEBUG(AR(printf("thrusters type 1\n")));
    shouldThrust = 1;
  }
  else if(backWall < 80 && frontWall > 250)
  {
    THRUSTDEBUG(AR(printf("thrusters type 2\n")));
    shouldThrust = 1;
  }
  else if(wall5 < 80 && frontWall > 250)
  {
    THRUSTDEBUG(AR(printf("thrusters type 4\n")));
    shouldThrust = 1;
  }
  else if(wall7 < 80 && frontWall > 250)
  {
    THRUSTDEBUG(AR(printf("thrusters type 5\n")));
    shouldThrust = 1;
  }
  else if(selfSpeed() < 3 && frontWall > 250 && wall1 > 100 && wall11 > 100)
  {
    shouldThrust = 1;
  }
  
  float turn = turnRules(closest, closest_angle, furthest_angle, selfSpeed());
  
  AR(printf("centroid: %f\n", turn));
  if((headingAimingDiff < 90 || headingAimingDiff > 270)&& backWall > 30)
  {
    fireShot(1);
  }
  if(turn < 1.0f)
  {
    // probabilistically turn left
    if(2 * fabs(turn - 0.5f) < (float)rand() / (float)RAND_MAX)
    {
      turnDir = -1;
    }
  }
  else if(turn > 1.0f)
  {
    // probabilistically turn right
    if(2 * fabs(turn - 1.5f) < (float)rand() / (float)RAND_MAX)
    {
      turnDir = 1;
    }
  }
  else
  {
    turnDir = 0;
  }
  if(selfSpeed() > 9 && headingTrackingDiff < 175)
  {
    TURNDEBUG(AR(printf("Turning reverse\n")));
    turnDir = -1;
  }
  //Speed too fast, turn to face against tracking direction
  else if(selfSpeed() > 9 && headingTrackingDiff > 185)
  {
    TURNDEBUG(AR(printf("Turning reverse\n")));
    turnDir = 1;
  }
  // if(turn < 0.75) turnDir = -1;
  // else if(turn < 1.25) turnDir = 0;
  // else turnDir = 1;

  if(closest > 90 && aimDir > 0 && headingAimingDiff < 180 && headingAimingDiff > 40)
  {
    TURNDEBUG(AR(printf("Aiming Roughly\n")));
    turnDir = 1;
  }
  else if(closest > 90 && aimDir > 0 && headingAimingDiff > 180 && headingAimingDiff < 320)
  {
    TURNDEBUG(AR(printf("Aiming Roughly\n")));
    turnDir = -1;
  }
  else if(closest > 90 && aimDir > 0)
  {
    TURNDEBUG(AR(printf("Aiming Accurately\n")))
    turnToDeg(aimDir);
    turnDir = 0;
  }

  if(shouldThrust) thrust(1);
  else thrust(0);
  if(turnDir == 1)
  {
    turnRight(1);
    turnLeft(0);
  }
  else if(turnDir == 0)
  {
    turnRight(0);
    turnLeft(0);
  }
  else if(turnDir == -1)
  {
    turnRight(0);
    turnLeft(1);
  }
  return 0;
}
int main(int argc, char *argv[]) {
  return start(argc, argv);
}

File: fuzzy/fuzzy.py
import os
import cma
import json
import math
import pickle
import numpy as np

# ---------- integration helpers ----------

def integrateLinearX(a, b, minx, maxx):
    """Integral of (a*x^2 + b*x) dx from minx to maxx."""
    if minx > maxx:
        return 0.0
    return a * (maxx**3 - minx**3) / 3.0 + b * (maxx**2 - minx**2) / 2.0

def integrateLinearXWithY(a, b, maxy):
    """Integral of (a*x^2 + b*x) dx between the x where y goes 0→maxy."""
    if a == 0.0:
        return 0.0  # mirrors safe behavior; C would divide by zero
    x1 = (maxy - b) / a
    x2 = -b / a
    lo, hi = (x2, x1) if x1 > x2 else (x1, x2)
    return integrateLinearX(a, b, lo, hi)

def integrateLinear(a, b, minx, maxx):
    """Integral of (a*x + b) dx from minx to maxx."""
    if minx > maxx:
        return 0.0
    return a * (maxx**2 - minx**2) / 2.0 + b * (maxx - minx)

def integrateLinearWithY(a, b, maxy):
    """Integral of (a*x + b) dx between the x where y goes 0→maxy."""
    if a == 0.0:
        return 0.0
    x1 = (maxy - b) / a
    x2 = -b / a
    lo, hi = (x2, x1) if x1 > x2 else (x1, x2)
    return integrateLinear(a, b, lo, hi)

def integrateBox(y, minx, maxx):
    if minx > maxx:
        return 0.0
    return y * (maxx - minx)

def integrateBoxX(y, minx, maxx):
    if minx > maxx:
        return 0.0
    return y * (maxx**2 - minx**2) / 2.0

# ---------- composite integrals (two ramps + flat) ----------

def integrateDoubleLinear(maxy, f1a, f1b, f2a, f2b):
    # Keep x1/x4 for parity with the C (x4 uses /f1a there and is unused)
    p1 = (maxy - f1b) / f1a if f1a != 0.0 else float('nan')
    p2 = (maxy - f2b) / f2a if f2a != 0.0 else float('nan')
    # assume p2 > p1
    return (integrateLinearWithY(f1a, f1b, maxy)
            + integrateBox(maxy, p1, p2)
            + integrateLinearWithY(f2a, f2b, maxy))

def integrateDoubleLinearX(maxy, f1a, f1b, f2a, f2b):
    p1 = (maxy - f1b) / f1a if f1a != 0.0 else float('nan')
    p2 = (maxy - f2b) / f2a if f2a != 0.0 else float('nan')
    # assume p2 > p1
    return (integrateLinearXWithY(f1a, f1b, maxy)
            + integrateBoxX(maxy, p1, p2)
            + integrateLinearXWithY(f2a, f2b, maxy))

# ---------- angles & membership functions ----------

def tanAngle(x):
    """Convert degrees x to tan(x/2), clamp to [-50, 50], special-case x==180 -> +50."""
    if x == 180:
        return 50.0
    t = math.tan(math.radians(x / 2.0))
    return max(min(t, 50.0), -50.0)

# Shared between ClosestAngle and FurthestAngle
def mem_Angle_Front(x, a=5, b=1.0):
    tanx = tanAngle(x)
    if tanx <= 0.0:
        return min(max(a*tanx+b, 0.0),1)
    else:
        return min(max(-a*tanx+b, 0.0),1)

def mem_Angle_Left(x, a=2,b=-1):
    tanx = tanAngle(x)
    return min(max(a*tanx+b, 0.0), 1.0)

def mem_Angle_Right(x, a=-2, b=-1):
    tanx = tanAngle(x)
    return min(max(a*tanx+b, 0.0), 1.0)

# Distance to closest wall
def mem_Wall_Safe(x, a=0.015, b=-5):
    return min(max(a*x+b, 0.0), 1.0)

def mem_Wall_Close(x, a=0.008, b=-1.2, c=-0.009, d=3.6):
    if a * x + b < 1.0:
        return min(max(a*x+b, 0.0), 1.0)
    else:
        return min(max(c*x+d, 0.0), 1.0)

def mem_Wall_Danger(x, a=-0.006, b=1.6):
    return min(max(a*x+b, 0.0), 1.0)

# ---------- defuzzification ----------

def centroidTurn(turnLeft, noTurn, turnRight):
    denum = (integrateDoubleLinear(turnLeft, 2, 0,  -2, 2)
             + integrateDoubleLinear(noTurn, 3, -2, -3, 4)
             + integrateDoubleLinear(turnRight, 2, -2, -2, 4))
    if denum == 0.0:
        return 1.0
    num = (integrateDoubleLinearX(turnLeft, 2, 0,  -2, 2)
           + integrateDoubleLinearX(noTurn, 3, -2, -3, 4)
           + integrateDoubleLinearX(turnRight, 2, -2, -2, 4))
    return num / denum

# ---------- fuzzy logic ops (avoid Python keywords) ----------

def f_and(a, b):
    return a if a < b else b

def f_or(a, b):
    return a if a > b else b

# ---------- rules ----------

def r1(closest, closestAngle, wd1, wd2, al1, al2):
    # If ClosestWall Danger and closestAngle Left then turn right
    return f_and(mem_Wall_Danger(closest, wd1, wd2), mem_Angle_Left(closestAngle, al1, al2))

def r2(closest, closestAngle, wd1, wd2, as1, as2):
    # If ClosestWall Danger and closestAngle Right then turn left
    return f_and(mem_Wall_Danger(closest, wd1, wd2), mem_Angle_Right(closestAngle, as1, as2))

def r3(closest, closestAngle, wc1, wc2, wc3, wc4, al1, al2):
    # If ClosestWall Close and closestAngle Left then turn right
    return f_and(mem_Wall_Close(closest, wc1, wc2, wc3, wc4), mem_Angle_Left(closestAngle, al1, al2))

def r4(closest, closestAngle, wc1, wc2, wc3, wc4, ar1, ar2):
    # If ClosestWall Close and closestAngle Right then turn left
    return f_and(mem_Wall_Close(closest, wc1, wc2, wc3, wc4), mem_Angle_Right(closestAngle, ar1, ar2))

def r5(closest, ws1, ws2):
    # If ClosestWall Safe Noturn
    return mem_Wall_Safe(closest, ws1, ws2)

def r6(furthestAngle, ar1, ar2):
    # If FurthestAngle Right turn Right
    return mem_Angle_Right(furthestAngle, ar1, ar2)

def r7(furthestAngle, al1, al2):
    # If FurthestAngle Left turn Left
    return mem_Angle_Left(furthestAngle, al1, al2)

def r8(furthestAngle, af1, af2):
    # If FurthestAngle Forward dont turn
    return mem_Angle_Front(furthestAngle, af1, af2)

def turnRules(closest, closestAngle, furthestAngle, chromosome):
    wd1 = chromosome[0]
    wd2 = chromosome[1]
    ar1 = chromosome[2]
    ar2 = chromosome[3]
    wc1 = chromosome[4]
    wc2 = chromosome[5]
    wc3 = chromosome[6]
    wc4 = chromosome[7]
    ws1 = chromosome[8]
    ws2 = chromosome[9]
    af1 = chromosome[10]
    af2 = chromosome[11]
    # rule activations
    # angle left is symmetrical to angle right
    turnLeft  = r2(closest, closestAngle, wd1, wd2, ar1, ar2) + r4(closest, closestAngle, wc1, wc2, wc3, wc4, ar1, ar2) + r7(furthestAngle, -ar1, ar2)
    noTurn    = r5(closest, ws1, ws2) + r8(furthestAngle, af1, af2)
    turnRight = r1(closest, closestAngle, wd1, wd2, -ar1, ar2) + r3(closest, closestAngle, wc1, wc2, wc3, wc4, ar1, ar2) + r6(furthestAngle, ar1, ar2)
    # defuzzify to crisp turn
    return centroidTurn(turnLeft, noTurn, turnRight)


import libpyAI as ai

from pathlib import Path
CKPT = Path("es.ckpt")

def load_or_create_es(x0, sigma0, opts):
    if CKPT.is_file():
        with CKPT.open("rb") as fh:
            es = pickle.load(fh)
        # keep logging, append to existing files
        es.logger = cma.CMADataLogger().register(es, append=True)
        print("[resume] loaded checkpoint:", CKPT)
    else:
        es = cma.CMAEvolutionStrategy(x0, sigma0, opts)
        es.logger = cma.CMADataLogger().register(es)
        print("[fresh] created new CMA-ES")
    return es

def save_ckpt(es, path=CKPT):
    tmp = Path(str(path) + ".tmp")
    with tmp.open("wb") as f:
        f.write(es.pickle_dumps())          # bytes with full optimizer state
    os.replace(tmp, path)

def AI_loop():
    if not hasattr(ai, "cmaes"):
        x0 = np.array([-0.006, 1.6, -2, -1, 0.008, -1.2, -0.009, 3.6, 0.015, -5, 5, 1], dtype=float)
        CMA_stds = [0.0012, 0.32, 0.4, 0.2, 0.0016, 0.24, 0.0018, 0.72, 0.003, 1.0, 1.0, 0.2]
        lower = [-0.02, -6, -6, -6, -0.02, -6, -0.02, -6, -0.02, -6, -6, -6]
        upper = [ 0.02,  6,  6,  6,  0.02,  6,  0.02,  6,  0.02,  6,  6,  6]
        opts = {"CMA_stds": CMA_stds, "bounds": [lower, upper]}
        es = load_or_create_es(x0,1.0, opts)
        setattr(ai, "cmaes", es)
        setattr(ai, "cmaes_candidates", es.ask())
        setattr(ai, "cmaes_candidates_count", len(ai.cmaes_candidates))
        setattr(ai, "cmaes_current", 0)
        setattr(ai, "cmaes_current_eval", 0)
        setattr(ai, "cmaes_current_fit", 0)
        setattr(ai, "cmaes_candidates_fit", [])
        setattr(ai, "cmaes_agent_alive", True)
        setattr(ai, "fitness", 1000000)

    if ai.selfAlive() == 0 and ai.cmaes_agent_alive: # agent just died
        print(f"Agent {ai.cmaes_current} Fitness: {ai.fitness}")
        if ai.cmaes_current_eval < 5:
            ai.cmaes_current_eval += 1
            ai.cmaes_current_fit += ai.fitness
        else:
            print(f"Agent {ai.cmaes_current} Avg Fitness: {ai.cmaes_current_fit / 5} Saving...")
            ai.cmaes_candidates_fit.append(ai.cmaes_current_fit / 5)
            ai.cmaes_current_fit = 0
            ai.cmaes_current_eval = 0
            ai.cmaes_current += 1
        if ai.cmaes_current == ai.cmaes_candidates_count:
            print(f"Generation {ai.cmaes.countiter} Done")
            ai.cmaes.tell(ai.cmaes_candidates, ai.cmaes_candidates_fit)
            setattr(ai, "cmaes_candidates", ai.cmaes.ask())
            setattr(ai, "cmaes_candidates_count", len(ai.cmaes_candidates))
            setattr(ai, "cmaes_current", 0)
            setattr(ai, "cmaes_candidates_fit", [])
            xbest = np.asarray(ai.cmaes.result.xbest)
            fbest = float(ai.cmaes.result.fbest)

            payload = {"xbest": xbest.tolist(), "fbest": fbest}
            with open("cma_best.json", "w") as f:
                json.dump(payload, f, indent=2)

            print(f"Best Chromosome: {xbest} Achieved: {fbest}")
            save_ckpt(ai.cmaes)
            setattr(ai, "cmaes_saved", True)
        setattr(ai, "fitness", 1000000)
        setattr(ai, "cmaes_agent_alive", False)
        
    if ai.selfAlive() == 1 and not ai.cmaes_agent_alive:
        setattr(ai, "cmaes_agent_alive", True)
        #print("New Agent")

    ai.fitness -= 1
    ai.setTurnSpeedDeg(20)
    aimDir = ai.aimdir(0)

    # Set variables
    heading = int(ai.selfHeadingDeg())
    tracking = int(ai.selfTrackingDeg())
    trackWall = ai.wallFeeler(500, tracking)

    # clockwise rotation around the ship
    frontWall = ai.wallFeeler(500, heading)
    wall5  = ai.wallFeeler(500, heading + 150)
    backWall = ai.wallFeeler(500, heading + 180)
    wall7  = ai.wallFeeler(500, heading + 210)

    shouldThrust = 0  # 0/1
    turnDir = 0       # right = 1, left = -1, none = 0

    # scan 360 degrees relative to heading
    furthest = 0.0
    furthest_angle = 0
    closest = 600.0
    closest_angle = 0
    for i in range(360):
        dist = ai.wallFeeler(500, heading + i)
        if dist > furthest:
            furthest = dist
            furthest_angle = i
        if dist < closest:
            closest = dist
            closest_angle = i
    headingTrackingDiff = int(heading + 360 - tracking) % 360
    headingAimingDiff = int(heading + 360 - aimDir) % 360  # kept for parity, unused below

    shotDanger = ai.shotAlert(0)
    if (shotDanger > 0 and shotDanger < 200 and frontWall > 200 and trackWall > 80):
        # print(f"avoiding shot: {shotDanger}")
        shouldThrust = 1
    elif (furthest_angle == 0) and (ai.selfSpeed() < 7) and (frontWall > 200):
        # print("propulsion")
        shouldThrust = 1
    elif (110 < headingTrackingDiff < 250) and (trackWall < 300) and (frontWall > 250):
        # print("thrusters type 1")
        shouldThrust = 1
    elif (backWall < 70) and (frontWall > 250):
        # print("thrusters type 2")
        shouldThrust = 1
    elif (85 < headingTrackingDiff < 275) and (trackWall < 100) and (frontWall > 200):
        # print("thrusters type 3")
        shouldThrust = 1
    elif (wall5 < 70) and (frontWall > 250):
        # print("thrusters type 4")
        shouldThrust = 1
    elif (wall7 < 70) and (frontWall > 250):
        # print("thrusters type 5")
        shouldThrust = 1
    elif ai.selfSpeed() < 5:
        shouldThrust = 1

    turn = turnRules(closest, closest_angle, furthest_angle, ai.cmaes_candidates[ai.cmaes_current])
    # print(f"centroid: {turn}")
    if turn < 0.75:
        turnDir = -1
    elif turn < 1.25:
        turnDir = 0
    else:
        turnDir = 1

    ai.thrust(1 if shouldThrust else 0)

    if turnDir == 1:
        ai.turnRight(1)
        ai.turnLeft(0)
    elif turnDir == 0:
        ai.turnRight(0)
        ai.turnLeft(0)
    elif turnDir == -1:
        ai.turnRight(0)
        ai.turnLeft(1)

    return 0



ai.start(AI_loop,["-name","Fuzzy","-join","localhost", ])